consensus/src/chained_bft/event_processor.rs

```diff
@@ -63,6 +64,59 @@ mod event_processor_test;
 #[path = "event_processor_fuzzing.rs"]
 pub mod event_processor_fuzzing;
 
+/// During the event of the node can't recover from local data, StartupSyncProcessor is responsible
+/// for processing the events carrying sync info and use the info to retrieve blocks from peers
+#[allow(dead_code)]
+pub struct StartupSyncProcessor<T> {
+    network: NetworkSender,
+    storage: Arc<dyn PersistentStorage<T>>,
+}
+
+#[allow(dead_code)]
+impl<T: Payload> StartupSyncProcessor<T> {
+    pub fn new(network: NetworkSender, storage: Arc<dyn PersistentStorage<T>>) -> Self {
+        StartupSyncProcessor { network, storage }
+    }
+
+    pub async fn process_proposal_msg(&mut self, proposal_msg: ProposalMsg<T>) -> RecoveryData<T> {
+        let author = proposal_msg.proposer();
+        let sync_info = proposal_msg.sync_info();
+        self.sync_up(&sync_info, author).await
+    }
+
+    pub async fn process_vote(&mut self, vote_msg: VoteMsg) -> RecoveryData<T> {
+        let author = vote_msg.vote().author();
+        let sync_info = vote_msg.sync_info();
+        self.sync_up(&sync_info, author).await
+    }
+
+    pub async fn process_sync_info_msg(
+        &mut self,
+        sync_info: SyncInfo,
+        peer: Author,
+    ) -> RecoveryData<T> {
+        debug!("Startup Sync received a sync info msg: {}", sync_info);
+        self.sync_up(&sync_info, peer).await
+    }
+
+    async fn sync_up(&mut self, sync_info: &SyncInfo, peer: Author) -> RecoveryData<T> {
+        let mut retriever = BlockRetriever::new(
+            self.network.clone(),
+            // Give a timeout of 1 hour to make sure there's enough time for trying to fetch from all peers
+            Instant::now() + Duration::new(3600, 0),
+            peer,
+        );
+        let (blocks, quorum_certs) =
+            BlockStore::fast_forward_sync(&sync_info.highest_commit_cert(), &mut retriever)
```

Comment:
we should check whether the sync_info is higher than local, otherwise the sync will fail

Reply:
Added check for `round` and `epoch`